File: tools/webgl-test-server.py
#!/usr/bin/env python3
"""
NativeBrowser WebGL Test Server

Serves Unity WebGL builds and provides test pages for verifying the
NativeBrowser WebGL plugin. Supports:

  - Serving a Unity WebGL build directory
  - Same-origin test page (for iframe JS execution tests)
  - Cross-origin test page (served on a second port)
  - Console log collection endpoint (POST /api/logs)
  - Health check endpoint (GET /api/health)

Usage:
    python tools/webgl-test-server.py [BUILD_DIR] [--port PORT] [--cross-origin-port PORT2]

    BUILD_DIR defaults to "webgl-build" in the working directory.

Examples:
    # Serve a WebGL build on port 8080 (cross-origin on 8081)
    python tools/webgl-test-server.py E:/test-project/WebGLBuild

    # Custom ports
    python tools/webgl-test-server.py E:/test-project/WebGLBuild --port 9000 --cross-origin-port 9001
"""

import argparse
import http.server
import json
import os
import socket
import sys
import threading
import time
from datetime import datetime
from pathlib import Path
from urllib.parse import parse_qs, urlparse

# ─── Collected Logs ────────────────────────────────────────────────────────────

_logs = []
_logs_lock = threading.Lock()


# ─── Test Pages ────────────────────────────────────────────────────────────────

SAME_ORIGIN_TEST_PAGE = """\
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>NativeBrowser Test — Same Origin</title>
  <style>
    body { font-family: system-ui, sans-serif; padding: 24px; background: #f5f5f5; }
    h1 { color: #333; }
    .card { background: #fff; border-radius: 8px; padding: 16px; margin: 12px 0;
            box-shadow: 0 1px 3px rgba(0,0,0,0.12); }
    button { padding: 8px 16px; margin: 4px; cursor: pointer; border: 1px solid #ccc;
             border-radius: 4px; background: #fff; }
    button:hover { background: #e8e8e8; }
    #log { background: #1e1e1e; color: #d4d4d4; padding: 12px; border-radius: 4px;
           font-family: monospace; font-size: 13px; max-height: 200px; overflow-y: auto;
           white-space: pre-wrap; }
  </style>
</head>
<body>
  <h1>NativeBrowser Test Page (Same Origin)</h1>

  <div class="card">
    <h3>PostMessage Test</h3>
    <button onclick="sendPostMessage()">Send PostMessage to Unity</button>
    <button onclick="listenPostMessage()">Start Listening</button>
  </div>

  <div class="card">
    <h3>JS Execution Test</h3>
    <p>This page exposes <code>window.testValue</code> and <code>window.getTestResult()</code>
    for verifying ExecuteJavaScript / InjectJavaScript from Unity.</p>
  </div>

  <div class="card">
    <h3>Console</h3>
    <div id="log"></div>
  </div>

  <script>
    window.testValue = "hello-from-test-page";

    window.getTestResult = function() {
      return JSON.stringify({ status: "ok", timestamp: Date.now() });
    };

    function log(msg) {
      var el = document.getElementById("log");
      el.textContent += "[" + new Date().toLocaleTimeString() + "] " + msg + "\\n";
      el.scrollTop = el.scrollHeight;
      // Also post to collection endpoint
      fetch("/api/logs", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ source: "test-page", message: msg, timestamp: Date.now() })
      }).catch(function() {});
    }

    function sendPostMessage() {
      window.parent.postMessage("test-message-from-iframe", "*");
      log("Sent postMessage: test-message-from-iframe");
    }

    function listenPostMessage() {
      window.addEventListener("message", function(e) {
        log("Received postMessage: " + (typeof e.data === "string" ? e.data : JSON.stringify(e.data)));
      });
      log("Listening for postMessages...");
    }

    log("Same-origin test page loaded. Origin: " + window.location.origin);
  </script>
</body>
</html>
"""

CROSS_ORIGIN_TEST_PAGE = """\
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>NativeBrowser Test — Cross Origin</title>
  <style>
    body { font-family: system-ui, sans-serif; padding: 24px; background: #fff3e0; }
    h1 { color: #e65100; }
    .card { background: #fff; border-radius: 8px; padding: 16px; margin: 12px 0;
            box-shadow: 0 1px 3px rgba(0,0,0,0.12); }
    button { padding: 8px 16px; margin: 4px; cursor: pointer; border: 1px solid #ccc;
             border-radius: 4px; background: #fff; }
    button:hover { background: #e8e8e8; }
  </style>
</head>
<body>
  <h1>NativeBrowser Test Page (Cross Origin)</h1>

  <div class="card">
    <h3>Cross-Origin PostMessage Test</h3>
    <p>This page is served on a different port to simulate cross-origin iframe behavior.</p>
    <button onclick="sendPostMessage()">Send PostMessage to Unity</button>
  </div>

  <div class="card">
    <h3>Notes</h3>
    <ul>
      <li>ExecuteJavaScript / InjectJavaScript should log cross-origin warnings</li>
      <li>PostMessage should still work cross-origin</li>
      <li>iframe load event should still fire</li>
    </ul>
  </div>

  <script>
    function sendPostMessage() {
      window.parent.postMessage("cross-origin-test-message", "*");
    }

    window.addEventListener("message", function(e) {
      console.log("[NativeBrowser CrossOrigin] Received:", e.data);
    });
  </script>
</body>
</html>
"""


# ─── Request Handler ───────────────────────────────────────────────────────────


class WebGLTestHandler(http.server.SimpleHTTPRequestHandler):
    """Extends SimpleHTTPRequestHandler with test page routes and API endpoints."""

    def __init__(self, *args, build_dir=None, cross_origin_port=None, **kwargs):
        self.build_dir = build_dir
        self.cross_origin_port = cross_origin_port
        super().__init__(*args, directory=build_dir, **kwargs)

    def do_GET(self):
        parsed = urlparse(self.path)
        path = parsed.path

        if path == "/api/health":
            self._json_response({"status": "ok", "time": datetime.now().isoformat()})
            return

        if path == "/api/logs":
            with _logs_lock:
                data = list(_logs)
            self._json_response({"logs": data, "count": len(data)})
            return

        if path == "/api/logs/clear":
            with _logs_lock:
                _logs.clear()
            self._json_response({"status": "cleared"})
            return

        if path == "/test/same-origin":
            self._html_response(SAME_ORIGIN_TEST_PAGE)
            return

        if path == "/test/cross-origin-info":
            self._json_response(
                {
                    "info": "Cross-origin test page is served on port "
                    + str(self.cross_origin_port),
                    "url": "http://localhost:"
                    + str(self.cross_origin_port)
                    + "/test/cross-origin",
                }
            )
            return

        # Default: serve static files from build directory
        super().do_GET()

    def do_POST(self):
        parsed = urlparse(self.path)
        path = parsed.path

        if path == "/api/logs":
            content_length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(content_length)
            try:
                entry = json.loads(body.decode("utf-8"))
                entry["_received"] = datetime.now().isoformat()
                with _logs_lock:
                    _logs.append(entry)
                    # Cap at 1000 entries
                    if len(_logs) > 1000:
                        _logs.pop(0)
                self._json_response({"status": "ok"})
            except json.JSONDecodeError:
                self._json_response({"error": "invalid json"}, status=400)
            return

        self.send_error(404)

    def _json_response(self, data, status=200):
        body = json.dumps(data, ensure_ascii=False).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(body)))
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        self.wfile.write(body)

    def _html_response(self, html, status=200):
        body = html.encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "text/html; charset=utf-8")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, format, *args):
        # Prefix with timestamp for readability
        sys.stderr.write(
            "[%s] %s\n" % (datetime.now().strftime("%H:%M:%S"), format % args)
        )

    def end_headers(self):
        # Add headers needed for Unity WebGL (SharedArrayBuffer requires these)
        self.send_header("Cross-Origin-Opener-Policy", "same-origin")
        self.send_header("Cross-Origin-Embedder-Policy", "require-corp")
        super().end_headers()


class CrossOriginHandler(http.server.BaseHTTPRequestHandler):
    """Minimal handler for the cross-origin test server."""

    def do_GET(self):
        parsed = urlparse(self.path)
        if parsed.path == "/test/cross-origin":
            body = CROSS_ORIGIN_TEST_PAGE.encode("utf-8")
            self.send_response(200)
            self.send_header("Content-Type", "text/html; charset=utf-8")
            self.send_header("Content-Length", str(len(body)))
            self.send_header("Access-Control-Allow-Origin", "*")
            self.end_headers()
            self.wfile.write(body)
        else:
            self.send_error(404)

    def log_message(self, format, *args):
        sys.stderr.write(
            "[%s] [cross-origin] %s\n"
            % (datetime.now().strftime("%H:%M:%S"), format % args)
        )


# ─── Server Setup ──────────────────────────────────────────────────────────────


def find_free_port(start=8080, end=9000):
    """Find an available port in the given range."""
    for port in range(start, end):
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            try:
                s.bind(("127.0.0.1", port))
                return port
            except OSError:
                continue
    raise RuntimeError(f"No free port found in range {start}-{end}")


def make_handler(build_dir, cross_origin_port):
    """Create a handler class with build_dir and cross_origin_port bound."""

    class BoundHandler(WebGLTestHandler):
        def __init__(self, *args, **kwargs):
            super().__init__(
                *args,
                build_dir=build_dir,
                cross_origin_port=cross_origin_port,
                **kwargs,
            )

    return BoundHandler


def run_server(build_dir, port, cross_origin_port):
    """Start both the main and cross-origin servers."""

    build_path = Path(build_dir).resolve()
    if not build_path.exists():
        print(f"Warning: Build directory does not exist yet: {build_path}")
        print(
            "Server will start but file serving will fail until the directory is created."
        )
        # Create it so SimpleHTTPRequestHandler doesn't crash
        build_path.mkdir(parents=True, exist_ok=True)

    # Main server
    handler_cls = make_handler(str(build_path), cross_origin_port)
    main_server = http.server.HTTPServer(("0.0.0.0", port), handler_cls)

    # Cross-origin server
    cross_server = http.server.HTTPServer(
        ("0.0.0.0", cross_origin_port), CrossOriginHandler
    )

    print("=" * 60)
    print("  NativeBrowser WebGL Test Server")
    print("=" * 60)
    print(f"  Build dir:       {build_path}")
    print(f"  Main server:     http://localhost:{port}")
    print(f"  Cross-origin:    http://localhost:{cross_origin_port}")
    print()
    print("  Endpoints:")
    print(f"    WebGL build:   http://localhost:{port}/")
    print(f"    Same-origin:   http://localhost:{port}/test/same-origin")
    print(f"    Cross-origin:  http://localhost:{cross_origin_port}/test/cross-origin")
    print(f"    Health check:  http://localhost:{port}/api/health")
    print(f"    Logs (GET):    http://localhost:{port}/api/logs")
    print(f"    Logs (POST):   http://localhost:{port}/api/logs")
    print(f"    Clear logs:    http://localhost:{port}/api/logs/clear")
    print("=" * 60)
    print("  Press Ctrl+C to stop")
    print()

    # Run cross-origin server in a daemon thread
    cross_thread = threading.Thread(target=cross_server.serve_forever, daemon=True)
    cross_thread.start()

    try:
        main_server.serve_forever()
    except KeyboardInterrupt:
        print("\nShutting down...")
        main_server.shutdown()
        cross_server.shutdown()


# ─── CLI ───────────────────────────────────────────────────────────────────────


def main():
    parser = argparse.ArgumentParser(
        description="NativeBrowser WebGL Test Server",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=__doc__,
    )
    parser.add_argument(
        "build_dir",
        nargs="?",
        default="webgl-build",
        help="Path to Unity WebGL build output directory (default: webgl-build)",
    )
    parser.add_argument(
        "--port",
        "-p",
        type=int,
        default=0,
        help="Main server port (default: auto-detect free port from 8080)",
    )
    parser.add_argument(
        "--cross-origin-port",
        "-c",
        type=int,
        default=0,
        help="Cross-origin test server port (default: main port + 1)",
    )

    args = parser.parse_args()

    # Auto-detect ports if not specified
    if args.port == 0:
        args.port = find_free_port(8080, 9000)

    if args.cross_origin_port == 0:
        args.cross_origin_port = find_free_port(args.port + 1, 9100)

    run_server(args.build_dir, args.port, args.cross_origin_port)


if __name__ == "__main__":
    main()

File: tools/webgl-test-server.mjs
#!/usr/bin/env node
/**
 * NativeBrowser WebGL Test Server (Node.js)
 *
 * Serves Unity WebGL builds with proper headers for:
 *   - Brotli-compressed files (.br) with correct Content-Encoding + Content-Type
 *   - COOP/COEP headers required for SharedArrayBuffer (Unity threading)
 *   - Same-origin and cross-origin test pages
 *   - Log collection API for automated testing
 *
 * Usage:
 *   node tools/webgl-test-server.mjs [BUILD_DIR] [--port PORT]
 *
 * Examples:
 *   node tools/webgl-test-server.mjs E:/webgl-build
 *   node tools/webgl-test-server.mjs E:/webgl-build --port 9000
 */

import http from "node:http";
import fs from "node:fs";
import path from "node:path";
import { URL } from "node:url";

// ─── Config ────────────────────────────────────────────────────────────────────

const args = process.argv.slice(2);
let buildDir = "webgl-build";
let port = 8080;

for (let i = 0; i < args.length; i++) {
  if (args[i] === "--port" || args[i] === "-p") {
    port = parseInt(args[++i], 10);
  } else if (!args[i].startsWith("-")) {
    buildDir = args[i];
  }
}

buildDir = path.resolve(buildDir);

// ─── MIME types ────────────────────────────────────────────────────────────────

const MIME_TYPES = {
  ".html": "text/html",
  ".htm": "text/html",
  ".js": "application/javascript",
  ".mjs": "application/javascript",
  ".css": "text/css",
  ".json": "application/json",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".svg": "image/svg+xml",
  ".ico": "image/x-icon",
  ".wasm": "application/wasm",
  ".data": "application/octet-stream",
  ".br": "application/octet-stream", // fallback, overridden below
  ".txt": "text/plain",
  ".xml": "application/xml",
  ".woff": "font/woff",
  ".woff2": "font/woff2",
  ".ttf": "font/ttf",
  ".otf": "font/otf",
};

/**
 * For Brotli files, determine MIME from the pre-compressed extension.
 * e.g. "foo.wasm.br" → application/wasm, "foo.js.br" → application/javascript
 */
function getMimeType(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  if (ext === ".br") {
    // Strip .br and check the real extension
    const innerExt = path.extname(filePath.slice(0, -3)).toLowerCase();
    return MIME_TYPES[innerExt] || "application/octet-stream";
  }
  return MIME_TYPES[ext] || "application/octet-stream";
}

// ─── Log Collection ────────────────────────────────────────────────────────────

const logs = [];

// ─── Test Pages ────────────────────────────────────────────────────────────────

const SAME_ORIGIN_TEST_PAGE = `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>NativeBrowser Test — Same Origin</title>
  <style>
    body { font-family: system-ui, sans-serif; padding: 24px; background: #f5f5f5; }
    h1 { color: #333; }
    .card { background: #fff; border-radius: 8px; padding: 16px; margin: 12px 0;
            box-shadow: 0 1px 3px rgba(0,0,0,0.12); }
    button { padding: 8px 16px; margin: 4px; cursor: pointer; border: 1px solid #ccc;
             border-radius: 4px; background: #fff; }
    button:hover { background: #e8e8e8; }
    #log { background: #1e1e1e; color: #d4d4d4; padding: 12px; border-radius: 4px;
           font-family: monospace; font-size: 13px; max-height: 200px; overflow-y: auto;
           white-space: pre-wrap; }
  </style>
</head>
<body>
  <h1>NativeBrowser Test Page (Same Origin)</h1>
  <div class="card">
    <h3>PostMessage Test</h3>
    <button onclick="sendPostMessage()">Send PostMessage to Unity</button>
    <button onclick="listenPostMessage()">Start Listening</button>
  </div>
  <div class="card">
    <h3>JS Execution Test</h3>
    <p>This page exposes <code>window.testValue</code> and <code>window.getTestResult()</code>
    for verifying ExecuteJavaScript / InjectJavaScript from Unity.</p>
  </div>
  <div class="card">
    <h3>Console</h3>
    <div id="log"></div>
  </div>
  <script>
    window.testValue = "hello-from-test-page";
    window.getTestResult = function() {
      return JSON.stringify({ status: "ok", timestamp: Date.now() });
    };
    function log(msg) {
      var el = document.getElementById("log");
      el.textContent += "[" + new Date().toLocaleTimeString() + "] " + msg + "\\n";
      el.scrollTop = el.scrollHeight;
      fetch("/api/logs", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ source: "test-page", message: msg, timestamp: Date.now() })
      }).catch(function() {});
    }
    function sendPostMessage() {
      window.parent.postMessage("test-message-from-iframe", "*");
      log("Sent postMessage: test-message-from-iframe");
    }
    function listenPostMessage() {
      window.addEventListener("message", function(e) {
        log("Received postMessage: " + (typeof e.data === "string" ? e.data : JSON.stringify(e.data)));
      });
      log("Listening for postMessages...");
    }
    log("Same-origin test page loaded. Origin: " + window.location.origin);
  </script>
</body>
</html>`;

// ─── Request Handler ───────────────────────────────────────────────────────────

function handleRequest(req, res) {
  const parsedUrl = new URL(req.url, `http://localhost:${port}`);
  const pathname = decodeURIComponent(parsedUrl.pathname);

  // Add COOP/COEP headers to ALL responses (required for Unity SharedArrayBuffer)
  res.setHeader("Cross-Origin-Opener-Policy", "same-origin");
  res.setHeader("Cross-Origin-Embedder-Policy", "require-corp");

  // ─── API Routes ──────────────────────────────────────────────────────────

  if (pathname === "/api/health") {
    jsonResponse(res, { status: "ok", time: new Date().toISOString() });
    return;
  }

  if (pathname === "/api/logs" && req.method === "GET") {
    jsonResponse(res, { logs, count: logs.length });
    return;
  }

  if (pathname === "/api/logs" && req.method === "POST") {
    let body = "";
    req.on("data", (chunk) => (body += chunk));
    req.on("end", () => {
      try {
        const entry = JSON.parse(body);
        entry._received = new Date().toISOString();
        logs.push(entry);
        if (logs.length > 1000) logs.shift();
        jsonResponse(res, { status: "ok" });
      } catch {
        jsonResponse(res, { error: "invalid json" }, 400);
      }
    });
    return;
  }

  if (pathname === "/api/logs/clear") {
    logs.length = 0;
    jsonResponse(res, { status: "cleared" });
    return;
  }

  if (pathname === "/test/same-origin") {
    htmlResponse(res, SAME_ORIGIN_TEST_PAGE);
    return;
  }

  // ─── Static File Serving ─────────────────────────────────────────────────

  let filePath = path.join(buildDir, pathname === "/" ? "index.html" : pathname);

  // Security: prevent directory traversal
  if (!path.resolve(filePath).startsWith(buildDir)) {
    res.writeHead(403);
    res.end("Forbidden");
    return;
  }

  // Try the exact path first
  if (!fs.existsSync(filePath)) {
    // If the file doesn't exist, check if there's a .br version
    // Unity WebGL loader will request e.g. "Build/foo.wasm" but the file is "Build/foo.wasm.br"
    if (fs.existsSync(filePath + ".br")) {
      filePath = filePath + ".br";
    } else {
      res.writeHead(404);
      res.end("Not Found");
      return;
    }
  }

  const stat = fs.statSync(filePath);
  if (stat.isDirectory()) {
    filePath = path.join(filePath, "index.html");
    if (!fs.existsSync(filePath)) {
      res.writeHead(404);
      res.end("Not Found");
      return;
    }
  }

  const mimeType = getMimeType(filePath);
  const headers = { "Content-Type": mimeType };

  // If serving a .br file, add Content-Encoding: br
  if (filePath.endsWith(".br")) {
    headers["Content-Encoding"] = "br";
  }

  headers["Content-Length"] = fs.statSync(filePath).size;

  res.writeHead(200, headers);
  fs.createReadStream(filePath).pipe(res);
}

function jsonResponse(res, data, status = 200) {
  const body = JSON.stringify(data);
  res.writeHead(status, {
    "Content-Type": "application/json",
    "Content-Length": Buffer.byteLength(body),
    "Access-Control-Allow-Origin": "*",
  });
  res.end(body);
}

function htmlResponse(res, html, status = 200) {
  res.writeHead(status, {
    "Content-Type": "text/html; charset=utf-8",
    "Content-Length": Buffer.byteLength(html),
  });
  res.end(html);
}

// ─── Start Server ──────────────────────────────────────────────────────────────

const server = http.createServer(handleRequest);

server.listen(port, "0.0.0.0", () => {
  console.log("=".repeat(60));
  console.log("  NativeBrowser WebGL Test Server (Node.js)");
  console.log("=".repeat(60));
  console.log(`  Build dir:       ${buildDir}`);
  console.log(`  Server:          http://localhost:${port}`);
  console.log();
  console.log("  Endpoints:");
  console.log(`    WebGL build:   http://localhost:${port}/`);
  console.log(`    Same-origin:   http://localhost:${port}/test/same-origin`);
  console.log(`    Health check:  http://localhost:${port}/api/health`);
  console.log(`    Logs (GET):    http://localhost:${port}/api/logs`);
  console.log(`    Logs (POST):   http://localhost:${port}/api/logs`);
  console.log(`    Clear logs:    http://localhost:${port}/api/logs/clear`);
  console.log("=".repeat(60));
  console.log("  Press Ctrl+C to stop");
  console.log();
});
